luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -0,0 +1,350 @@
+				shieldData.downtimeReset = 0
+				shieldData.shieldEnabled = true
+				spSetUnitShieldRechargeDelay(shieldUnitID, shieldData.shieldWeaponNumber, 0)
+
+				-- This section is to allow slower moving projectiles already inside the shield when it comes back online to damage units within the radius.
+				local x, y, z = spGetUnitPosition(shieldUnitID)
+				local radius = shieldData.radius-25
+				local xmin = x - radius
+				local xmax = x + radius
+				local zmin = z - radius
+				local zmax = z + radius
+				local projectiles = spGetProjectilesInRectangle(xmin, zmin, xmax, zmax)
+				for _, projectileID in ipairs(projectiles) do
+					projectileShieldHitCache[projectileID] = true
+				end
+			end
+		end
+	end
+
+	if frame % 90 == 0 then
+		shieldUnitsTotalCount = 0
+		shieldUnitIndex = {}
+
+		for shieldUnitID, shieldData in pairs(shieldUnitsData) do
+			shieldUnitsTotalCount = shieldUnitsTotalCount + 1
+			shieldUnitIndex[shieldUnitsTotalCount] = shieldUnitID
+		end
+
+		shieldCheckChunkSize = math.max(math.ceil(shieldUnitsTotalCount / 10), 1)
+
+		if lastShieldCheckChunkNumber > #shieldUnitIndex then
+			lastShieldCheckChunkNumber = 1
+		end
+
+		shieldCheckEndIndex = math.min(lastShieldCheckChunkNumber + shieldCheckChunkSize - 1, #shieldUnitIndex)
+	end
+
+	for i = lastShieldCheckChunkNumber, shieldCheckEndIndex do
+		local shieldUnitID = shieldUnitIndex[i]
+		local shieldData = shieldUnitsData[shieldUnitID]
+
+		if shieldData then
+			if shieldData.shieldCoverageChecked == false then
+				if shieldData.shieldEnabled == true then
+					setCoveredUnits(shieldUnitID)
+				else
+					removeCoveredUnits(shieldUnitID)
+				end
+			end
+
+			shieldData.shieldCoverageChecked = false
+		end
+	end
+
+	lastShieldCheckChunkNumber = shieldCheckEndIndex + 1
+
+	if lastShieldCheckChunkNumber > #shieldUnitIndex then
+		lastShieldCheckChunkNumber = 1
+	end
+end
+
+function gadget:UnitPreDamaged(unitID, unitDefID, unitTeam, damage, paralyzer, weaponDefID, projectileID, attackerID, attackerDefID, attackerTeam)
+	if not AOEWeaponDefIDs[weaponDefID] or projectileShieldHitCache[projectileID] then
+		return damage
+	end
+
+	if shieldNegatesDamageCheck(unitID, unitTeam, attackerID, attackerTeam) then
+		return 0, 0
+	else
+		return damage
+	end
+end
+
+function gadget:ShieldPreDamaged(proID, proOwnerID, shieldWeaponNum, shieldUnitID, bounceProjectile, beamEmitterWeaponNum, beamEmitterUnitID, startX, startY, startZ, hitX, hitY, hitZ)
+	local shieldData = shieldUnitsData[shieldUnitID]
+	local weaponDefID
+
+	if not shieldData or not shieldData.shieldEnabled then
+		return true
+	end
+
+	shieldData.shieldWeaponNumber = shieldWeaponNum
+
+	-- proID isn't nil if hitscan weapons are used, it's actually -1.
+	if proID > -1 then
+		weaponDefID = projectileDefIDCache[proID]
+
+		if not weaponDefID then
+			-- Because flame projectiles for some reason don't live long enough to reference from the cache table
+			weaponDefID = spGetProjectileDefID(proID)
+		end
```

Comment:
```suggestion
		weaponDefID = projectileDefIDCache[proID] or spGetProjectileDefID(proID)
```
Flame weapons may not have `Script.SetWatchWeapon` set, which would cause them to not hit the callins that do the caching.